token/lib/token.js

```diff
@@ -100,6 +110,148 @@ class Token {
+    if (alreadyPaused) {
+      throw new Error('Token is already paused')
+    }
+    const tokenOwner = await contract.methods.owner().call()
+    if (tokenOwner.toLowerCase() != sender.toLowerCase()) {
+      throw new Error(`Sender ${sender} is not owner of token contract (${tokenOwner})`)
+    }
+
+    const transaction = contract.methods.pause()
+    await this.sendTransaction(networkId, transaction, { from: sender })
+    if (await contract.methods.paused().call() !== true) {
+      throw new Error('Token should be paused but is not')
+    }
+  }
+
+  /**
+   * Unpauses transfers and approvals of tokens.
+   * @param {string} networkId - Ethereum network ID.
+   */
+  async unpause(networkId) {
+    const contract = await this.contract(networkId)
+    const sender = this.defaultAccount(networkId)
+
+    // Pre-contract call validations.
+    const paused = await contract.methods.paused().call()
+    if (!paused) {
+      throw new Error('Token is already unpaused')
+    }
+    const tokenOwner = await contract.methods.owner().call()
+    if (tokenOwner.toLowerCase() != sender.toLowerCase()) {
+      throw new Error(`Sender ${sender} is not owner of token contract (${tokenOwner})`)
+    }
+
+    const transaction = contract.methods.unpause()
+    await this.sendTransaction(networkId, transaction, { from: sender })
+    if (await contract.methods.paused().call() !== false) {
+      throw new Error('Token should be unpaused but is not')
+    }
+  }
+
+  // TODO: refactor into separate base class, to support other contracts such
+  // as the marketplace contract
+  /**
+   * Sends an Ethereum transaction.
+   * @param {string} networkId - Ethereum network ID.
+   * @param {transaction} transaction - These are returned by contract.methods.MyMethod()
+   * @param {Object} opts - Options to be sent along with the transaction.
+   * @returns {Object} - Transaction receipt.
+   */
+  async sendTransaction(networkId, transaction, opts = {}) {
+    // TODO: support multisig wallets
+
+    const web3 = this.web3(networkId)
+
+    let transactionHash
+    if (!opts.gas) {
+      opts.gas = await transaction.estimateGas({ from: opts.from })
+      this.vlog('estimated gas:', opts.gas)
+    }
+
+    // Send the transaction and grab the transaction hash when it's available.
+    this.vlog('sending transaction')
+    transaction.send(opts)
+      .on('transactionHash', (hash) => {
+        transactionHash = hash
+        this.vlog('transaction hash:', transactionHash)
+      })
+
+    // Poll for the transaction receipt, with an exponential backoff. This works
+    // around some strange interactions between web3.js and some web3 providers.
+    // For example, this issue sometimes prevents transaction receipts from
+    // being returned when simply calling
+    // `await contract.methods.MyMethod(...).send(...)`:
+    //
+    // https://github.com/INFURA/infura/issues/95
+    const maxSleep = 15 * 60 * 1000 // 15 minutes, in milliseconds
```

Comment:
Wow. 15min is a long time ? I think on average blocks are mined every 15sec.
But I guess it does not really hurt to wait.

Reply:
Yes, good point. Fixed.